blockchain/peer_test.go

```diff
@@ -0,0 +1,241 @@
+func TestPeerIncrPending(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+
+	peer.incrPending()
+	assert.NotNil(t, peer.recvMonitor)
+	assert.NotNil(t, peer.timeout)
+	assert.Equal(t, int32(1), peer.numPending)
+
+	peer.incrPending()
+	assert.NotNil(t, peer.recvMonitor)
+	assert.NotNil(t, peer.timeout)
+	assert.Equal(t, int32(2), peer.numPending)
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+}
+
+func TestPeerDecrPending(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+
+	// panic if numPending is 0 and try to decrement it
+	assert.Panics(t, func() { peer.decrPending(10) })
+
+	// decrement to zero
+	peer.incrPending()
+	peer.decrPending(10)
+	assert.Equal(t, int32(0), peer.numPending)
+	// make sure timer is not running
+	checkByStoppingPeerTimer(t, peer, false)
+
+	// decrement to non zero
+	peer.incrPending()
+	peer.incrPending()
+	peer.decrPending(10)
+	assert.Equal(t, int32(1), peer.numPending)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+}
+
+func TestPeerCanBeRemovedDueToExpiration(t *testing.T) {
+	minRecvRate = int64(100) // 100 bytes/sec exponential moving average
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		errFunc: errFunc,
+		logger:  log.TestingLogger(),
+	}
+
+	peerTimeout = time.Millisecond
+	peer.incrPending()
+	time.Sleep(2 * time.Millisecond)
+	// timer expired, should be able to remove peer
+	peerTestMtx.Lock()
+	assert.Equal(t, errNoPeerResponse, lastErr)
+	peerTestMtx.Unlock()
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+
+}
+
+func TestPeerCanBeRemovedDueToLowSpeed(t *testing.T) {
+	minRecvRate = int64(100) // 100 bytes/sec exponential moving average
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		errFunc: errFunc,
+		logger:  log.TestingLogger(),
+	}
+
+	peerTimeout = time.Second
+	peerSampleRate = 0
+	peerWindowSize = 0
+
+	peer.incrPending()
+	peer.numPending = 100
+
+	// monitor starts with a higher rEMA (~ 2*minRecvRate), wait for it to go down
+	time.Sleep(900 * time.Millisecond)
+
+	// normal peer - send a bit more than 100 byes/sec, > 10 byes/100msec, check peer is not considered slow
+	for i := 0; i < 10; i++ {
+		peer.decrPending(11)
+		time.Sleep(100 * time.Millisecond)
+		require.Nil(t, peer.isGood())
+	}
+
+	// slow peer - send a bit less than 10 byes/100msec
+	for i := 0; i < 10; i++ {
+		peer.decrPending(9)
+		time.Sleep(100 * time.Millisecond)
+	}
+	// check peer is considered slow
+	assert.Equal(t, errSlowPeer, peer.isGood())
+
+}
+
+func TestPeerCleanup(t *testing.T) {
```

Comment:
Should probably be removed.

Reply:
I was thinking to keep it around and do a commit dedicated to cleanup :)
The question I have is what do we do once we are in finished state. We leave quite a few stale things around right now and I plan to clean. We do stop timers I think, but peers are left around, etc. IMO we should scan the pool when finished and verify it is in the correct state and remove peers.